### DISCUSSION TITLE
Seeed Studio XIAO nRF52840 Sense support for memory->pwm output

### DISCUSSION THREAD

**Discussion by BobaTrek** [pschatzmann/arduino-audio-tools#976] [Q&A] [answered]
I am attempting to play a .wav file from memory of the XIAO nRF52840 Sense to a PWM output pin.

The XIAO board is nearly the same as the Nano 33 BLE, but also includes an IMU which I require.  Here is a link to the board:
[https://www.seeedstudio.com/Seeed-XIAO-BLE-Sense-nRF52840-p-5253.html](url)

I have tried compiling the "streams-memory_wav-pwm.ino" using both the Arduino Mbed Nano 33 BLE board and the Seeed XIAO nRF52840 Sense Mbed board and get the same error on both attempts:

```
BaseConverter.h:117:45: error: conversion from 'int' to 'audio_tools::int24_t' is ambiguous
```

Here is the complete output of the compile attempt:

```
In file included from d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools.h:56:0,
                 from D:\Robert\DEV\Arduino\Bob\streams-memory_wav-pwm\streams-memory_wav-pwm.ino:11:
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools/BaseConverter.h: In instantiation of 'audio_tools::ConverterAutoCenterT<T>::ConverterAutoCenterT(int) [with T = audio_tools::int24_t]':
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools/BaseConverter.h:206:73:   required from here
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools/BaseConverter.h:117:45: error: conversion from 'int' to 'audio_tools::int24_t' is ambiguous
         ConverterAutoCenterT(int channels=2){
                                             ^
In file included from d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioConfig.h:694:0,
                 from d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools.h:51,
                 from D:\Robert\DEV\Arduino\Bob\streams-memory_wav-pwm\streams-memory_wav-pwm.ino:11:
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioBasic/Int24.h:42:3: note: candidate: audio_tools::int24_t::int24_t(float)
   int24_t(const float in) {
   ^~~~~~~
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioBasic/Int24.h:34:3: note: candidate: audio_tools::int24_t::int24_t(int64_t)
   int24_t(const int64_t in) {
   ^~~~~~~
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioBasic/Int24.h:30:3: note: candidate: audio_tools::int24_t::int24_t(int32_t)
   int24_t(const int32_t in) {
   ^~~~~~~
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioBasic/Int24.h:26:3: note: candidate: audio_tools::int24_t::int24_t(int16_t)
   int24_t(const int16_t in) {
   ^~~~~~~
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioBasic/Int24.h:22:3: note: candidate: audio_tools::int24_t::int24_t(void*)
   int24_t(void *ptr) {
   ^~~~~~~
In file included from d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTimer/AudioTimerBase.h:4:0,
                 from d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTimer/AudioTimer.h:10,
                 from d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools.h:52,
                 from D:\Robert\DEV\Arduino\Bob\streams-memory_wav-pwm\streams-memory_wav-pwm.ino:11:
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools/AudioTypes.h: In instantiation of 'static T audio_tools::NumberConverter::clip(int64_t) [with T = audio_tools::int24_t; int64_t = long long int]':
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools/AudioTypes.h:277:29:   required from 'static ToT audio_tools::NumberConverter::convert(FromT) [with FromT = short int; ToT = audio_tools::int24_t]'
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools/AudioStreamsConverter.h:292:57:   required from 'size_t audio_tools::NumberFormatConverterStreamT<TFrom, TTo>::write(const uint8_t*, size_t) [with TFrom = short int; TTo = audio_tools::int24_t; size_t = unsigned int; uint8_t = unsigned char]'
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools/AudioStreamsConverter.h:420:64:   required from here
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools/AudioTypes.h:268:25: error: conversion from 'int' to 'audio_tools::int24_t' is ambiguous
                 return -mv;
                         ^~
In file included from d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioConfig.h:694:0,
                 from d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioTools.h:51,
                 from D:\Robert\DEV\Arduino\Bob\streams-memory_wav-pwm\streams-memory_wav-pwm.ino:11:
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioBasic/Int24.h:42:3: note: candidate: audio_tools::int24_t::int24_t(float)
   int24_t(const float in) {
   ^~~~~~~
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioBasic/Int24.h:34:3: note: candidate: audio_tools::int24_t::int24_t(int64_t)
   int24_t(const int64_t in) {
   ^~~~~~~
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioBasic/Int24.h:30:3: note: candidate: audio_tools::int24_t::int24_t(int32_t)
   int24_t(const int32_t in) {
   ^~~~~~~
d:\Robert\DEV\Arduino\libraries\arduino-audio-tools\src/AudioBasic/Int24.h:26:3: note: candidate: audio_tools::int24_t::int24_t(int16_t)
   int24_t(const int16_t in) {
   ^~~~~~~

exit status 1

Compilation error: exit status 1
```

I have searched the discussions and the only related thread I found was this one where someone was having difficulty compiling for the XIAO nRF82540 as well, but was having a different error with the I2S portion.
[https://github.com/pschatzmann/arduino-audio-tools/discussions/141](url)

In that discussion, @pschatzmann said he committed some changes to support ARDUINO_SEEED_XIAO_NRF52840_SENSE.  It ran into I2S problems, but as I am using PWM, I hope to be able to get this to work.

Any help would be appreciated!

EDIT:  As a test, I switched to the XIAO_ESP32S3 board target, and it compiles without error for that board.  I then tried the Arduino Nano RP2040 Connect (Mbed) which was unable to locate I2S.h, so did not compile.

**Comment by pschatzmann** [comment, marked as answer]
Double check the AudioConfig.h and make sure that the relevant processor contains a #define USE_INT24_FROM_INT
If you use the latest commit, this should already be there.

**Reply by BobaTrek** [reply]
It did not.  So I added the following and it is compiling successfully now.  Here is where I inserted it in AudioConfig.h (last line below):

```//------ NANO33BLE ----------
#if defined(ARDUINO_SEEED_XIAO_NRF52840_SENSE) || defined(ARDUINO_ARDUINO_NANO33BLE)
#define USE_NANO33BLE 
#define USE_I2S
#define USE_PWM
#define USE_TYPETRAITS
#define USE_EFFECTS_SUITE
#define USE_TIMER
#define USE_INT24_FROM_INT  // Added per  @pschatzmann
```

I will download and test on the target board tomorrow.  Thanks for the assistance!

BTW:  How do I determine which pin on the board will have the PWM signal on it?  I see this line of code:
```
#define PIN_PWM_START 6
```
On the back of the PCB board there are 10 I/O pins silkscreened 0 through 10.  Presumably the mapping is 1:1.  The pinout from the XIAO documentation is attached below.
![pinout2](https://github.com/pschatzmann/arduino-audio-tools/assets/34546883/7a124b27-7b41-4cd8-b967-49373e20c386)

**Reply by BobaTrek** [reply]
Perhaps my next few steps might help someone else trying to use the XIAO nRF52840 for PWM audio output.

Using the schematic for the XIAO nRF52840 sense, I was able to find the pin mapping:
![XIAO nRF52840 pin mapping](https://github.com/pschatzmann/arduino-audio-tools/assets/34546883/b4b0e67e-1a49-4ff9-88c1-b795204a460e)
So from here you can see that D0 - D10 do not map 1:1 with the chip's physical pins.  Pad D6 (net name P1.11_TX_D6) is actually mapped to the chips port 1.11 which is Pin 43 (port#1*32 + bit#11) for the way Arduino assigns pin numbers.   In the AudioConfig,h file I changed the following:
```
#define PIN_PWM_START 43  \\ Pad D6 is chip port 1.11
```
That did not work though.

I also tried using D6 to specify the pad, which also did not work.
```
#define PIN_PWM_START D6  \\ Pad D6
```

So I decided to try the D5 pad on the board which is mapped to port P0.05 of the chip, or "Pin 5" in Arduino terminology.  In the AudioConfig,h file I changed the following:
```
#define PIN_PWM_START 5  \\ Use pad D5, chip port 0.05, or Arduino pin 5 instead
```
That worked and I started getting PWM audio out on pad D5.

This also worked for routing to pad D5:
```
#define PIN_PWM_START D5  \\ Pad D5
```

The audio is extremely low volume, so next I am going to hook up an I2S amplifier as the audio sink and see how that goes.

**Comment by pschatzmann** [comment]
Not sure where the confusion comes from: All pins are can/should be specified with the Arduino conventions.

To make a sketch more speaking I propose to use an explicit method to set the pins by calling setPins() on the config object.

As far as the volume is concerned you have all the tools to analyse the situation: Maybe the recorded volume in your wav is just too low and you can consider to amplify it with the VolumeStream. 

If you output a sine wave that oscillates between negative 32768 through positive 32767 you should see a sine wave oscillating between 0V and 3.3V on the pwm pins.

**Reply by BobaTrek** [reply]
The only confusion I had was XIAO nRF52840 specific in that D6, which was used as the default in the PWM sample code, for some reason does not function on the XIAO nRF52840.  But D5 did function.  Although I didn't look too far, I don't see anything in the documentation that would preclude PWM from being sent to D6.  So I thought I would put this info in this thread just in case someone else ran into the same thing.

I did change my code to the following per your suggestion:

```
  Pins pwmPins = Pins();
  pwmPins.push_back(5);
  config.setPins(pwmPins);
```

So everything is going quite well actually.  This audio-tools library is intense!

**Comment by pschatzmann** [comment]
The following should work as well:
```
  Pins pwmPins = {5};
  config.setPins(pwmPins);
```

**Reply by BobaTrek** [reply]
I tried that form first and it didn't work, so I tried the more elaborate version.

When I put
```
  Pins pwmPins = {5};
  config.setPins(pwmPins);
```
I get
```
[I] PWMAudioMBED.h : 77 - PWM Pin: 6
```

When I put
```
  Pins pwmPins = Pins();
  pwmPins.push_back(5);
```
I get
```
[I] PWMAudioMBED.h : 77 - PWM Pin: 5
```

But what was really strange was that when I used the Pins pwmPins = {5} form, the pin reported in the Info message from line 77 would actually be whatever the prior runs value was.  So not what was in the current PIN_PWM_START, but rather the value from a prior run.  But the pwmPins.push_back(5) form always worked.  Very strange.

I do not know if it is related, but I also always get the audio rate and audio channels reported as values of 0:
```
[I] D:\Robert\DEV\Arduino\Bob\streams-memory_wav-pwm\streams-memory_wav-pwm.ino : 51 - The audio rate from the wav file is 0
[I] D:\Robert\DEV\Arduino\Bob\streams-memory_wav-pwm\streams-memory_wav-pwm.ino : 52 - The channels from the wav file is 0
```

**Reply by pschatzmann** [reply]
Yes, this might be the reason: This does only work when the PWMStream contains the correct audio info. 
I don't see how this would be possible in my original code. Maybe your wav file has an invalid header ?

Can you share your code and the log with level info ?

After a code review I identified a potential issue. I committed the correction...

**Reply by BobaTrek** [reply]
I started off with the streams-memory_wav-pwm.ino example in the audiotools repository and made two main changes.  I added delay(6000) to give me time to open and clear the serial monitor between uploads, and I added the code to set pin 5 as the pwm output pin.

The alice.h and knghtsng.h are unmodified from the repository.  I was unable to use knghtsng.h because it overflowed the available 1 Mbyte of memory on the XIAO nRF52840.

I was unable to find the "Partition Scheme Huge APP" option referred to in the README.md for the XIAO nRF52840 mbed-enabled board.  The XIAO also has 2 Mbytes of QSPI flash which I hope to use to store the audio in the final version.

Here is the code:
```
/**
 * @file stream-memory_wav-pwm.ino
 * @author Phil Schatzmann
 * @brief decode WAV stream and output to PWM pins
 * @version 0.1
 * @date 2021-01-24
 * 
 * @copyright Copyright (c) 2021
 */

#include "AudioTools.h"
//#include "knghtsng.h"
#include "alice.h"


//Data Flow: MemoryStream -> EncodedAudioStream  -> PWMAudioOutput
//Use 8000 for alice_wav and 11025 for knghtsng_wav
AudioInfo info(8000, 1, 16);
//MemoryStream wav(knghtsng_wav, knghtsng_wav_len);
MemoryStream wav(alice_wav, alice_wav_len);
PWMAudioOutput pwm;          // PWM output 
EncodedAudioStream out(&pwm, new WAVDecoder()); // Decoder stream
StreamCopy copier(out, wav);    // copy in to out

void setup(){
  Serial.begin(115200);
  delay(6000);  // Had to increase so that Arduino serial monitor would be ready
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
 
  auto config = pwm.defaultConfig();
  config.copyFrom(info);
  //Pins pwmPins = {5};  // Does not work
  Pins pwmPins = Pins();
  pwmPins.push_back(5);
  config.setPins(pwmPins);

  wav.begin();
  out.begin();

  pwm.begin(config);
}

void loop(){
  if (wav) {
    copier.copy();
  } else {
    // after we are done we just print some info form the wav file
    auto info = out.audioInfo();
    LOGI("The audio rate from the wav file is %d", info.sample_rate);
    LOGI("The channels from the wav file is %d", info.channels);

    // restart from the beginning
    Serial.println("Restarting...");
    delay(5000);
    out.begin();   // indicate that we process the WAV header
    wav.begin();       // reset actual position to 0
    pwm.begin();       // reset counters 
  }
}
```

And here is the serial monitor output:
```
[I] PWMAudioBase.h : 124 - decimation: 1
[I] PWMAudioBase.h : 136 - ->Allocating new buffer 4 * 1024 bytes
[I] Buffers.h : 321 - resize: 4096
[I] PWMAudioBase.h : 84 - sample_rate: 8000
[I] PWMAudioBase.h : 85 - channels: 1
[I] PWMAudioBase.h : 86 - bits_per_sample: 16
[I] PWMAudioBase.h : 87 - buffer_size: 1024
[I] PWMAudioBase.h : 88 - buffer_count: 4
[I] PWMAudioBase.h : 89 - pwm_frequency: 30000
[I] PWMAudioBase.h : 90 - resolution: 8
[I] PWMAudioMBED.h : 77 - PWM Pin: 5
[I] PWMAudioBase.h : 153 - ->Buffer available: 0
[I] PWMAudioBase.h : 154 - ->Buffer available for write: 4096
[I] PWMAudioBase.h : 155 - ->is_timer_started: false 
[I] CodecWAV.h : 57 - WAVHeader::write: 1024 -> 44 -> 44
[I] CodecWAV.h : 63 - WAVHeader::begin: 44
[I] CodecWAV.h : 219 - WAVHeader sound_pos: 44
[I] CodecWAV.h : 220 - WAVHeader channels: 1 
[I] CodecWAV.h : 221 - WAVHeader bits_per_sample: 16
[I] CodecWAV.h : 222 - WAVHeader sample_rate: 8000 
[I] CodecWAV.h : 223 - WAVHeader format: 1
[I] CodecWAV.h : 367 - WAV sample_rate: 8000
[I] CodecWAV.h : 368 - WAV data_length: 2147418104
[I] CodecWAV.h : 369 - WAV is_streamed: 1
[I] CodecWAV.h : 371 - WAV is_valid: true
[I] CodecWAV.h : 393 - isValid: true
[I] AudioPWM.h : 39 - virtual void audio_tools::PWMAudioOutput::setAudioInfo(audio_tools::AudioInfo)
[I] CodecWAV.h : 396 - WAVDecoder writing first sound data
[I] StreamCopy.h : 305 - try write - 2 (open 0 bytes) 
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 2 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 139 - StreamCopy::copy 766 -> 766 -> 766 bytes - in 1 hops
[I] D:\Robert\DEV\Arduino\Bob\streams-memory_wav-pwm\streams-memory_wav-pwm.ino : 49 - The audio rate from the wav file is 0
[I] D:\Robert\DEV\Arduino\Bob\streams-memory_wav-pwm\streams-memory_wav-pwm.ino : 50 - The channels from the wav file is 0
Restarting...
```

**Comment by pschatzmann** [comment]
I assume that the log is from the version before my correction. The log confirms that the WAV has the correct information, so I pretty confident that my correction will help

The partition scheme is an ESP32 thing, so this will not help you.

In your case, I suggest that you look into other codecs: ADPCM or WAV with ADPCM might be an interesting option,  or you could use an 8bits binary file instead. I doubt that there is enough memory for using mp3.

**Reply by BobaTrek** [reply]
I just downloaded the most recent version with the following results:

1)  The AudioConfig.h now contains the statement "#define USE_INT24_FROM_INT" for the ARDUINO_SEEED_XIAO_NRF52840_SENSE.  Thanks!

2)  I am still getting 0 reported in the two Info statements:
```
[I] D:\Robert\DEV\Arduino\Bob\streams-memory_wav-pwm\streams-memory_wav-pwm.ino : 49 - The audio rate from the wav file is 0
[I] D:\Robert\DEV\Arduino\Bob\streams-memory_wav-pwm\streams-memory_wav-pwm.ino : 50 - The channels from the wav file is 0
```

3)  The following still does not work:
```
  Pins pwmPins = {5};  // Does not work
  config.setPins(pwmPins);
```
but the following still works:
```
  //Pins pwmPins = {5};  // Does not work
  Pins pwmPins = Pins();
  pwmPins.push_back(5);
  config.setPins(pwmPins);
```

**Reply by pschatzmann** [reply]
Just made a test and for me it is working like a charm:
```

  auto config = pwm.defaultConfig();
  config.copyFrom(info);
  //config.resolution = 8;  // must be between 8 and 11 -> drives pwm frequency (8 is default)
  // alternative 1
  //config.start_pin = 3;
  // alternative 2: defines pins and channels
  int pins[] = {3};
  config.setPins(pins); 
  pwm.begin(config);
```
Is giving in the log
```
[I] PWMAudioMBED.h : 77 - PWM Pin: 2
```

**Reply by BobaTrek** [reply]
I have the same results.  Option 1 fails while Options  2, 3, and 4 succeed:
```
  // Option 1 does not work - calls setPins(Pins &pins)
  // Pins pwmPins = {5};

  // Option 2 works - calls setPins(Pins &pins)
  //Pins pwmPins = Pins();
  //pwmPins.push_back(5);

  // Option 3 works - calls setPins(T (&a) [N])
  int pwmPins[] = {5};

  // Option 4 works:
  //config.start_pin = 5;

  config.setPins(pwmPins);
```
I don't understand why Option 1 fails while Option 2 succeeds.  They both call setPins(Pins &pins).  The only difference is the way the Pins vector is created.

Pins is defined as "using Pins = Vector<int>;"
What is the subtle difference between declaring pwmPins using Option 1 verses Option 2?

**Reply by pschatzmann** [reply]
I think I found it. The last line in the set setPins(Pins) should not be there

**Reply by BobaTrek** [reply]
I removed the line "pins_data = pins;" and Option 1 is still failing for me.

**Comment by pschatzmann** [comment]
I activated the USE_INITIALIZER_LIST in the config. 
This should do the trick...

**Reply by BobaTrek** [reply]
Yes, all of the setPins options work now.  Thanks!
